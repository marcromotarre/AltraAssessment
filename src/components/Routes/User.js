import React from 'react';
import { useSelector } from 'react-redux';
import { useParams } from '@reach/router';

import { default as UserContainer } from '../User/User.js';
import SectionBanner from '../Banners/SectionBanner';
import DataManager from '../DataManager.js';

const User = () => {
  let { id } = useParams();
  id = parseInt(id);
  const users = useSelector((state) => state.town.users);
  const user = users.find((user) => user.id === id);
  const friends = users.filter(({ name } = u) => user.friends.includes(name));
  return (
    <div data-testid="UserSite">
      <SectionBanner />
      <DataManager>
        <UserContainer {...user} friends={friends} />
      </DataManager>
    </div>
  );
};

export default User;
