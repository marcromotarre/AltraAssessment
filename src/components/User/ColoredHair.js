import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import HairSVG from './../../svg/hair.svg';

const HairIcon = styled(HairSVG)`
  path {
    stroke: ${(props) => props.strokecolor};
    fill: ${(props) => props.color};
  }
`;

const ColoredHair = ({ color, strokecolor, width, height }) => {
  return (
    <HairIcon
      color={color}
      strokecolor={strokecolor}
      width={width}
      height={height}
    />
  );
};

ColoredHair.propTypes = {
  color: PropTypes.string,
  strokecolor: PropTypes.string,
  width: PropTypes.number,
};

ColoredHair.defaultProps = {
  strokecolor: 'rgba(0,0,0,0)',
  color: '#000',
};

export default ColoredHair;
