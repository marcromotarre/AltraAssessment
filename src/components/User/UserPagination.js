import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const UserPaginationContainer = styled.button`
  margin: 0;
  border: 0;
  width: 100%;
  padding: 10px 30px 10px 10px;
  height: 40px;
  border-bottom: 1px solid #e3e3e3;
  background-color: #f8f8f8;
  display: flex;
  justify-content: center;
  align-items: center;
  label {
    font-family: 'HelveticaNeue-Bold';
    color: #707070;
    font-size: 16px;
  }
`;

const UserPagination = ({ text, onClick }) => {
  const onClickContainer = () => {
    onClick();
  };
  return (
    <UserPaginationContainer onClick={onClickContainer}>
      <label>{text}</label>
    </UserPaginationContainer>
  );
};

UserPagination.propTypes = {
  text: PropTypes.string,
  onClick: PropTypes.func,
};

UserPagination.defaultProps = {
  text: 'Load more users',
  onClick: () => {},
};

export default UserPagination;
