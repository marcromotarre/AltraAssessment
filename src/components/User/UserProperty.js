import React from 'react';
import styled from 'styled-components';

const UserPropertyContainer = styled.div`
  display: flex;
  justify-content: space-evenly;
  align-items: center;
  width: 100%;
  label {
    font-family: 'HelveticaNeue-Thin';
    color: #707070;
    font-size: 12px;
  }

  .big {
    font-size: 22px;
  }
`;

const UserProperty = ({ children }) => {
  return (
    <UserPropertyContainer data-testid="user-property">
      {children}
    </UserPropertyContainer>
  );
};

export default UserProperty;
