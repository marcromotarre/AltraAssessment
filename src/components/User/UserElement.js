import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import HeightSVG from './../../svg/height.svg';
import WeightSVG from './../../svg/weight.svg';
import CandleSVG from './../../svg/candle.svg';
import FriendsSVG from './../../svg/friends.svg';
import ProfessionsSVG from './../../svg/professions.svg';

import UserProperty from './../User/UserProperty.js';
import ColoredHair from './ColoredHair';

const UserContainer = styled.section`
  padding: 10px 30px 10px 10px;
  display: grid;
  grid-template-columns: 58px auto;
  column-gap: 10px;
  border-bottom: 1px solid #e3e3e3;
  background-color: #f8f8f8;
`;

const UserImage = styled.img`
  border-radius: 50%;
  height: 58px;
  width: 58px;
  justify-self: center;
  align-self: center;
`;
const UserInfo = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
`;

const UserName = styled.label`
  width: 100%;
  justify-content: start;
  align-items: center;
  font-family: 'HelveticaNeue-Bold';
  color: #707070;
  font-size: 16px;
`;

const UserProperties = styled.div`
  grid-area: m-properties;
  width: 100%;
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const UserElement = ({
  name,
  thumbnail,
  age,
  weight,
  height,
  hair_color,
  professions,
  friends,
}) => {
  return (
    <UserContainer>
      <UserImage data-testid="user-image" src={thumbnail} />
      <UserInfo>
        <UserName data-testid="user-name">{name}</UserName>
        <UserProperties data-testid="user-properties">
          <UserProperty>
            <CandleSVG width={14.63} /> <label>{age}</label>
          </UserProperty>
          <UserProperty data-testid="user-height">
            <HeightSVG height={17.7} />
            <label>
              {isNaN(height) ? height : Math.floor(height).toFixed(0)}
            </label>
          </UserProperty>
          <UserProperty data-testid="user-width">
            <WeightSVG width={17.7} />
            <label>
              {isNaN(weight) ? weight : Math.floor(weight).toFixed(0)}
            </label>
          </UserProperty>
          <UserProperty>
            <ColoredHair color={hair_color} width={15.26} />
          </UserProperty>
          <UserProperty>
            <ProfessionsSVG width={14.34} />
            <label>{professions.length}</label>
          </UserProperty>
          <UserProperty>
            <FriendsSVG width={15.26} /> <label>{friends.length}</label>
          </UserProperty>
        </UserProperties>
      </UserInfo>
    </UserContainer>
  );
};

UserElement.propTypes = {
  name: PropTypes.string.isRequired,
  thumbnail: PropTypes.string.isRequired,
  age: PropTypes.number.isRequired,
  weight: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  hair_color: PropTypes.string.isRequired,
  professions: PropTypes.arrayOf(PropTypes.string).isRequired,
  friends: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default UserElement;
