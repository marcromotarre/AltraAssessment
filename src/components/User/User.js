import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import HeightSVG from './../../svg/height.svg';
import WeightSVG from './../../svg/weight.svg';
import AgeSVG from './../../svg/candle.svg';
import FriendsSVG from './../../svg/friends.svg';
import ProfessionsSVG from './../../svg/professions.svg';

import SectionTitle from './../SectionTitle.js';

import UserProperty from './../User/UserProperty.js';
import Profession from '../Profession';
import UserElement from './UserElement';
import { Link } from '@reach/router';
import ColoredHair from './ColoredHair';

export const UserContainer = styled.section`
  display: grid;
  grid-template-columns: 100%;
  row-gap: 25px;

  a {
    text-decoration: none;
  }
`;

export const UserImage = styled.img`
  border-radius: 50%;
  height: 137px;
  width: 137px;
  justify-self: center;
  align-self: center;
`;

export const UserName = styled.label`
  justify-self: center;
  align-self: center;
  font-family: 'HelveticaNeue-Bold';
  color: #707070;
  font-size: 24px;
`;

export const UserInfo = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
`;

export const UserProperties = styled.div`
  width: 80%;
  justify-self: center;
  align-self: center;
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

export const Section = styled.section`
  display: grid;
  grid-template-columns: 100%;
  row-gap: 15px;
`;

export const ProfessionsList = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  flex-wrap: wrap;
  padding: 0px 40px;
  a {
    > section {
      margin: 5px;
    }
  }
`;

const User = ({
  name,
  thumbnail,
  age,
  weight,
  height,
  hair_color,
  professions = [],
  friends = [],
}) => {
  return (
    <UserContainer data-testid="actual-user-container">
      <UserImage data-testid="actual-user-image" src={thumbnail} />
      <UserName data-testid="actual-user-name">{name}</UserName>
      <UserProperties data-testid="actual-user-properties">
        <UserProperty>
          <AgeSVG width={28.39} /> <label className="big">{age}</label>
        </UserProperty>
        <UserProperty>
          <HeightSVG height={33.36} />
          <label className="big">{Math.floor(height).toFixed(0)}</label>
        </UserProperty>
        <UserProperty>
          <WeightSVG width={33.36} />
          <label className="big">{Math.floor(weight).toFixed(0)}</label>
        </UserProperty>
        <UserProperty>
          <ColoredHair color={hair_color} width={39.5} />
        </UserProperty>
      </UserProperties>
      {professions.length > 0 && (
        <Section data-testid="actual-user-professions">
          <SectionTitle name="Professions" number={professions.length}>
            <ProfessionsSVG width={25.05} />
          </SectionTitle>
          <ProfessionsList>
            {professions.map((profession, index) => (
              <Link key={index} to={`/profession/${profession}`}>
                <Profession key={index} name={profession} />
              </Link>
            ))}
          </ProfessionsList>
        </Section>
      )}
      {friends.length > 0 && (
        <Section data-testid="actual-user-friends">
          <SectionTitle name="Friends" number={friends.length}>
            <FriendsSVG width={31.5} />
          </SectionTitle>
          <div>
            {friends.map((friend, index) => (
              <Link key={index} to={`/user/${friend.id}`}>
                <UserElement {...friend} />
              </Link>
            ))}
          </div>
        </Section>
      )}
    </UserContainer>
  );
};

export default User;
