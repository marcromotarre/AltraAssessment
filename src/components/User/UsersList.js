import React, { useState } from 'react';
import { Link } from '@reach/router';
import styled from 'styled-components';
import PropTypes from 'prop-types';

import UserElement from './UserElement.js';
import UserPagination from './UserPagination.js';

export const UsersListContainer = styled.div`
  a {
    text-decoration: none;
  }
`;

const UsersList = ({ users, usersPerPage }) => {
  const [page, setPage] = useState(1);

  const loadMoreUsers = () => {
    setPage(page + 1);
  };

  const moreUsers = usersPerPage * page >= users.length ? false : true;

  return (
    <UsersListContainer>
      {users &&
        users
          .filter((a, index) => index < usersPerPage * page)
          .map((user) => (
            <Link key={user.id} to={`/user/${user.id}`}>
              <UserElement {...user} />
            </Link>
          ))}
      {moreUsers && <UserPagination onClick={loadMoreUsers} />}
    </UsersListContainer>
  );
};

UsersList.propTypes = {
  users: PropTypes.array,
  usersPerPage: PropTypes.number,
};

UsersList.defaultProps = {
  users: [],
  usersPerPage: Number.POSITIVE_INFINITY,
};

export default UsersList;
