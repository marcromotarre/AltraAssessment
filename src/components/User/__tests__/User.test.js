import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { shallow } from 'enzyme';
import User from '../User';

/** this component should:
 *  - render all user detailed data
 *    - including:
 *      - all professions user data are links to professions route
 *      - all clickable friends user data to user routes
 */

describe('User Element Component tests', () => {
  const friend = {
    id: 'friend_id',
    name: 'user name',
    thumbnail: 'imagepath',
    age: 20,
    weight: 70.86543,
    height: 150.83749,
    hair_color: 'Green',
    professions: [
      'profession_1',
      'profession_2',
      'profession_3',
      'profession_4',
    ],
    friends: ['friend1', 'friend2', 'friend3'],
  };

  const props = {
    name: 'user name',
    thumbnail: 'imagepath',
    age: 20,
    weight: 70.86543,
    height: 150.33749,
    hair_color: 'Green',
    professions: ['profession_1', 'profession_2'],
    friends: [friend],
  };

  const propsNoFriendNoProfessions = {
    name: 'user name',
    thumbnail: 'imagepath',
    age: 20,
    weight: 70.86543,
    height: 150.33749,
    hair_color: 'Green',
    professions: [],
    friends: [],
  };

  it('render UserElement', () => {
    const tree = renderer.create(<User {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render container and its children', () => {
    const { getByTestId } = render(<User {...props} />);
    const container = getByTestId('actual-user-container');
    expect(container.children.length).toBe(5);
  });

  it('should render user image', () => {
    const { getByTestId } = render(<User {...props} />);
    const userImage = getByTestId('actual-user-image');
    expect(userImage.getAttribute('src')).toBe('imagepath');
  });

  it('should render user name', () => {
    const { getByTestId } = render(<User {...props} />);
    const userName = getByTestId('actual-user-name');
    expect(userName.innerHTML).toBe(props.name);
  });

  it('should render user properties (age. width, height, hair_color)', () => {
    const { getByTestId } = render(<User {...props} />);
    const userProperties = getByTestId('actual-user-properties');
    expect(userProperties.children.length).toBe(4);

    const age = userProperties.children[0];
    expect(age.children.length).toBe(2);
    expect(age.querySelector('svg').getAttribute('width')).toBe('28.39');
    expect(age.querySelector('label').innerHTML).toBe(`${props.age}`);
    expect(age.querySelector('label').getAttribute('class')).toBe('big');

    const height = userProperties.children[1];
    expect(height.children.length).toBe(2);
    expect(height.querySelector('svg').getAttribute('height')).toBe('33.36');
    expect(height.querySelector('label').innerHTML).toBe('150');
    expect(height.querySelector('label').getAttribute('class')).toBe('big');

    const weight = userProperties.children[2];
    expect(weight.children.length).toBe(2);
    expect(weight.querySelector('svg').getAttribute('width')).toBe('33.36');
    expect(weight.querySelector('label').innerHTML).toBe('70');
    expect(weight.querySelector('label').getAttribute('class')).toBe('big');

    const hairColor = userProperties.children[3];
    expect(hairColor.children.length).toBe(1);
    expect(hairColor.querySelector('svg').getAttribute('width')).toBe('39.5');
    expect(hairColor.querySelector('svg').getAttribute('color')).toBe(
      props.hair_color
    );
  });

  it('should not render professions if user has no professions & should not render friends if user has no friends', () => {
    const { getByTestId } = render(<User {...propsNoFriendNoProfessions} />);
    const container = getByTestId('actual-user-container');
    expect(container.children.length).toBe(3);
  });

  it('should render professions', () => {
    const wrapper = shallow(<User {...props} />);
    const professions = wrapper.find('[data-testid="actual-user-professions"]');
    const sectionTitle = professions.childAt(0);
    expect(sectionTitle.prop('name')).toBe('Professions');
    expect(sectionTitle.prop('number')).toBe(props.professions.length);
    expect(sectionTitle.childAt(0).prop('width')).toBe(25.05);
    const professionsList = professions.childAt(1);
    expect(professionsList.find('Link').length).toBe(props.professions.length);
    expect(professionsList.childAt(0).prop('to')).toBe(
      `/profession/${props.professions[0]}`
    );
    expect(professionsList.childAt(1).prop('to')).toBe(
      `/profession/${props.professions[1]}`
    );
  });

  it('should render friends', () => {
    const wrapper = shallow(<User {...props} />);
    const friends = wrapper.find('[data-testid="actual-user-friends"]');
    const sectionTitle = friends.childAt(0);
    expect(sectionTitle.prop('name')).toBe('Friends');
    expect(sectionTitle.prop('number')).toBe(props.friends.length);
    expect(sectionTitle.childAt(0).prop('width')).toBe(31.5);
    const friendsList = friends.childAt(1);
    expect(friendsList.find('Link').length).toBe(props.friends.length);
    expect(friendsList.childAt(0).prop('to')).toBe(
      `/user/${props.friends[0].id}`
    );
  });
});
