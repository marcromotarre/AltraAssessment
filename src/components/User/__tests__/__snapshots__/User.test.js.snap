// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Element Component tests render UserElement 1`] = `
<section
  className="sc-fzpans bVyTxg"
  data-testid="actual-user-container"
>
  <img
    className="sc-fzplWN dysndN"
    data-testid="actual-user-image"
    src="imagepath"
  />
  <label
    className="sc-fznyAO hFMGcT"
    data-testid="actual-user-name"
  >
    user name
  </label>
  <div
    className="sc-fznZeY klUTXZ"
    data-testid="actual-user-properties"
  >
    <div
      className="sc-AxhCb gfcCsk"
      data-testid="user-property"
    >
      <svg
        width={28.39}
      />
       
      <label
        className="big"
      >
        20
      </label>
    </div>
    <div
      className="sc-AxhCb gfcCsk"
      data-testid="user-property"
    >
      <svg
        height={33.36}
      />
      <label
        className="big"
      >
        150
      </label>
    </div>
    <div
      className="sc-AxhCb gfcCsk"
      data-testid="user-property"
    >
      <svg
        width={33.36}
      />
      <label
        className="big"
      >
        70
      </label>
    </div>
    <div
      className="sc-AxhCb gfcCsk"
      data-testid="user-property"
    >
      <svg
        className="sc-AxgMl grTrYX"
        color="Green"
        width={39.5}
      />
    </div>
  </div>
  <section
    className="sc-fzokOt hTyMhG"
    data-testid="actual-user-professions"
  >
    <section
      className="sc-AxjAm eDLRgU"
    >
      <section
        className="sc-AxirZ gXuGhx"
      >
        <svg
          width={25.05}
        />
        <label>
          Professions
        </label>
      </section>
      <label
        className="sc-AxiKw jbzapM"
      >
        <label>
          2
        </label>
      </label>
    </section>
    <div
      className="sc-fzqBZW ipBCnN"
    >
      <a
        href="/profession/profession_1"
        onClick={[Function]}
      >
        <section
          className="sc-AxhUy gEDYiZ"
        >
          <label>
            profession_1
          </label>
        </section>
      </a>
      <a
        href="/profession/profession_2"
        onClick={[Function]}
      >
        <section
          className="sc-AxhUy gEDYiZ"
        >
          <label>
            profession_2
          </label>
        </section>
      </a>
    </div>
  </section>
  <section
    className="sc-fzokOt hTyMhG"
    data-testid="actual-user-friends"
  >
    <section
      className="sc-AxjAm eDLRgU"
    >
      <section
        className="sc-AxirZ gXuGhx"
      >
        <svg
          width={31.5}
        />
        <label>
          Friends
        </label>
      </section>
      <label
        className="sc-AxiKw jbzapM"
      >
        <label>
          1
        </label>
      </label>
    </section>
    <div>
      <a
        href="/user/friend_id"
        onClick={[Function]}
      >
        <section
          className="sc-AxheI cuTdVv"
        >
          <img
            className="sc-Axmtr hcomXe"
            data-testid="user-image"
            src="imagepath"
          />
          <div
            className="sc-AxmLO cIzxDa"
          >
            <label
              className="sc-fzozJi gpkgCw"
              data-testid="user-name"
            >
              user name
            </label>
            <div
              className="sc-fzoLsD livjYH"
              data-testid="user-properties"
            >
              <div
                className="sc-AxhCb gfcCsk"
                data-testid="user-property"
              >
                <svg
                  width={14.63}
                />
                 
                <label>
                  20
                </label>
              </div>
              <div
                className="sc-AxhCb gfcCsk"
                data-testid="user-property"
              >
                <svg
                  height={17.7}
                />
                <label>
                  150
                </label>
              </div>
              <div
                className="sc-AxhCb gfcCsk"
                data-testid="user-property"
              >
                <svg
                  width={17.7}
                />
                <label>
                  70
                </label>
              </div>
              <div
                className="sc-AxhCb gfcCsk"
                data-testid="user-property"
              >
                <svg
                  className="sc-AxgMl grTrYX"
                  color="Green"
                  width={15.26}
                />
              </div>
              <div
                className="sc-AxhCb gfcCsk"
                data-testid="user-property"
              >
                <svg
                  width={14.34}
                />
                <label>
                  4
                </label>
              </div>
              <div
                className="sc-AxhCb gfcCsk"
                data-testid="user-property"
              >
                <svg
                  width={15.26}
                />
                 
                <label>
                  3
                </label>
              </div>
            </div>
          </div>
        </section>
      </a>
    </div>
  </section>
</section>
`;
